### PR TITLE
Postgres integration

### PR DESCRIPTION
**Pull request by elongl**

### CODE REVIEW

File: macros/edr/alerts/anomaly_detection_description.sql [hunk 1]
```diff
@@ -9,21 +9,21 @@
 {% endmacro %}
 
 {% macro freshness_description() %}
-    'Last update was at ' || anomalous_value || ', ' || abs(round(metric_value/3600,2)) || ' hours ago. Usually the table is updated within ' || abs(round(training_avg/3600,2)) || ' hours.'
+    'Last update was at ' || anomalous_value || ', ' || abs(round({{ elementary.cast_as_numeric('metric_value/3600') }}, 2)) || ' hours ago. Usually the table is updated within ' || abs(round({{ elementary.cast_as_numeric('metric_value/3600') }}, 2)) || ' hours.'
```
Comment (haritamar) [inline macros/edr/alerts/anomaly_detection_description.sql:12]: Guessing it's a mistake - the second should be `training_avg` right?

File: macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql [hunk 2]
```diff
@@ -10,7 +10,7 @@
             edr_bucket_start,
             edr_bucket_end,
             1 as joiner
-          from ({{ elementary.complete_buckets_cte(time_bucket) }})
+          from ({{ elementary.complete_buckets_cte(time_bucket) }}) rs
```
Comment (haritamar) [inline macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql:13]: What does "rs" stand for?

File: macros/edr/metadata_collection/get_tables_from_information_schema.sql [hunk 3]
```diff
@@ -68,3 +67,28 @@
...
+    )
+
+    select
+        {{ elementary.full_table_name() }} as full_table_name,
+        upper(database_name || '.' || schema_name) as full_schema_name,
+        database_name,
+        schema_name,
+        table_name
+    from information_schema_tables
+    )
```
Comment (haritamar) [inline macros/edr/metadata_collection/get_tables_from_information_schema.sql:92]: Why is the query wrapped with parantheses?

File: macros/edr/tests/test_utils/get_compiled_test_id.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{% macro get_compiled_test_id() %}
+    {% set test_node = model %}
+    {% set test_hash = test_node.unique_id.split(".")[-1] %}
+    {% set test_name = test_node.name %}
+    {% do return("{}_{}".format(test_name, test_hash)) %}
```
Comment (haritamar) [inline macros/edr/tests/test_utils/get_compiled_test_id.sql:5]: Can add context on why this was needed?
I'm slightly worried about adding another type of ID for tests to our code, it's confusing :)

File: macros/utils/data_types/cast_column.sql [hunk 5]
```diff
@@ -6,6 +6,10 @@
     cast({{ column }} as {{ elementary.type_float() }})
 {%- endmacro -%}
 
+{%- macro cast_as_numeric(column) -%}
+    cast({{ column }} as {{ elementary.type_numeric() }})
+{%- endmacro -%}
```
Comment (haritamar) [inline macros/utils/data_types/cast_column.sql:11]: Can add context on why casting to `type_numeric` was needed?

File: macros/utils/table_operations/create_or_replace.sql [hunk 6]
```diff
@@ -4,17 +4,24 @@
...
+    {% do dbt.drop_relation_if_exists(relation) %}
+    {% do run_query(dbt.create_table_as(temporary, relation, sql_query)) %}
     {% do adapter.commit() %}
 {% endmacro %}
 
+{% macro postgres__create_or_replace(temporary, relation, sql_query) %}
+    {% do run_query("BEGIN") %}
+    {% do dbt.drop_relation_if_exists(relation) %}
+    {% do run_query(dbt.create_table_as(temporary, relation, sql_query)) %}
+    {% do run_query("COMMIT") %}
```
Comment (haritamar) [inline macros/utils/table_operations/create_or_replace.sql:20]: I'm wondering if the Postgres implementation specifically should be in a transaction if the rest are not, WDYT?

File: macros/utils/table_operations/table_name_with_suffix.sql [hunk 7]
```diff
@@ -2,7 +2,7 @@
     {% set relation_max_name_length = elementary.get_relation_max_name_length()  %}
     {% if relation_max_name_length %}
         {% set suffix_length = suffix | length %}
-        {% set table_name_with_suffix = table_name[:relation_max_name_length - suffix_length] ~ suffix %}
+        {% set table_name_with_suffix = table_name[-(relation_max_name_length - suffix_length):] ~ suffix %}
```
Comment (haritamar) [inline macros/utils/table_operations/table_name_with_suffix.sql:5]: Why?

File: macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql (hunk 2 above)
Comment (elongl) [reply macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql:13]: Result-set.
I was also unaware at first, the naming is a community contribution.
Should we change it to `results`?

File: macros/utils/data_types/cast_column.sql (hunk 5 above)
Comment (elongl) [reply macros/utils/data_types/cast_column.sql:11]: The `round()` function of Postgres expects this data type as its first argument.

```sql
Database Error
  function round(double precision, integer) does not exist
  LINE 16:                 else round(execution_time, 1)
                                ^
  HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

File: macros/utils/table_operations/create_or_replace.sql (hunk 6 above)
Comment (elongl) [reply macros/utils/table_operations/create_or_replace.sql:20]: I don't have an alternative really.
Calling `adapter.commit()` fails because it says that I haven't started a transaction.
And not doing so just doesn't commit.

The docs says that it's okay to do so however.
image

File: macros/edr/metadata_collection/get_tables_from_information_schema.sql (hunk 3 above)
Comment (elongl) [reply macros/edr/metadata_collection/get_tables_from_information_schema.sql]: I think the contributor just followed the others' pattern.
There doesn't seem to be a need in those parentheses.
Removed.

File: macros/edr/tests/test_utils/get_compiled_test_id.sql (hunk 4 above)
Comment (elongl) [reply macros/edr/tests/test_utils/get_compiled_test_id.sql:5]: The problem is conflicting names for Elementary's test tables.
The current table names are `{flattened_test.name}__metrics` or `{flattened_test.name}__anomaly_scores`.
Here's an example of the problem with the current situation:
```py
ipdb> t1 = ctx['graph']['nodes']['test.elementary_integration_tests.elementary_source_freshness_anomalies_training_any_type_column_anoma
lies_training_.5733415eda']
ipdb> t2 = ctx['graph']['nodes']['test.elementary_integration_tests.elementary_source_freshness_anomalies_training_string_column_anomali
es_training_.4640cc98af']
ipdb> t1['name']
'elementary_source_freshness_anomalies_training_any_type_column_anomalies_training_'
ipdb> t2['name']
'elementary_source_freshness_anomalies_training_string_column_anomalies_training_'
ipdb> prefix = '__anomaly_scores'
ipdb> max_length = 63
ipdb> t1["name"][:max_length - len(prefix)] + prefix
'elementary_source_freshness_anomalies_training___anomaly_scores'
ipdb> t2["name"][:max_length - len(prefix)] + prefix
'elementary_source_freshness_anomalies_training___anomaly_scores'
```

As you can see, two different tests (`t1`, `t2`) yield the same table name and therefore override each other and error upon running `dbt test`.

The solution is to include the test's hash into the table so that it's unique.
In addition to that, the table's name is now truncated from the end rather than from the start in order to include the hash. Alternatively we could've placed the hash first but then we also need to add dummy characters because the hash might start with digits, but table names need to start with letters.

Please let me know what you think.
Thanks.

File: macros/utils/table_operations/table_name_with_suffix.sql (hunk 7 above)
Comment (elongl) [reply macros/utils/table_operations/table_name_with_suffix.sql:5]: Explanation [here](https://github.com/elementary-data/dbt-data-reliability/pull/230/files#r1070637099).

File: .github/workflows/test-all-warehouses.yml [hunk 8]
```diff
@@ -12,7 +12,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        warehouse-type: [ snowflake, bigquery, redshift, databricks ]
+        warehouse-type: [ postgres, snowflake, bigquery, redshift, databricks ]
```
Comment (Maayan-s) [inline .github/workflows/test-all-warehouses.yml:15]: Regardless of this PR, should we add Spark here?

File: integration_tests/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# dbt-data-reliability dbt Package Tests
```
Comment (Maayan-s) [inline integration_tests/README.md:1]: Cool!
I imagine this is for contributors? 
Should we update here as well? 
https://docs.elementary-data.com/general/contributions

File: macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql (hunk 2 above)
Comment (elongl) [reply macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql]: Decided to rename it.

Comment (Maayan-s) [reply macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql]: He called it result set -> rs . 
I would prefer to use a name that describes what the result set is, makes the code more readable

File: integration_tests/README.md (hunk 9 above)
Comment (elongl) [reply integration_tests/README.md:1]: Generally yes, but at the moment I don't think so because we have a task this sprint to improve our testing infrastructure in which hopefully we'll have something more systematic than our current script.

File: macros/edr/metadata_collection/get_columns_from_information_schema.sql [hunk 10]
```diff
@@ -33,4 +32,19 @@
...
+    {%- set database_name, schema_name = schema_tuple %}
+
+    select
+        upper(table_catalog || '.' || table_schema || '.' || table_name) as full_table_name,
+        upper(table_catalog) as database_name,
+        upper(table_schema) as schema_name,
+        upper(table_name) as table_name,
+        upper(column_name) as column_name,
+        data_type
+    from information_schema.columns
```
Comment (Maayan-s) [inline macros/edr/metadata_collection/get_columns_from_information_schema.sql:47]: Can we change to relay on dbt here like on the default query? feels safer

File: macros/edr/metadata_collection/get_tables_from_information_schema.sql [hunk 11]
```diff
@@ -7,7 +7,7 @@
     {%- set database_name, schema_name = schema_tuple %}
     {% set schema_relation = api.Relation.create(database=database_name, schema=schema_name).without_identifier() %}
 
-    (with information_schema_tables as (
+    with information_schema_tables as (
```
Comment (Maayan-s) [inline macros/edr/metadata_collection/get_tables_from_information_schema.sql:10]: Why was this removed? 
I think it was there for a reason

File: macros/edr/metadata_collection/get_tables_from_information_schema.sql [hunk 12]
```diff
@@ -39,15 +39,12 @@
...
-    )
-
 {% endmacro %}
 
-
 {% macro redshift__get_tables_from_information_schema(schema_tuple) %}
     {%- set database_name, schema_name = schema_tuple %}
 
-    (with information_schema_tables as (
+    with information_schema_tables as (
```
Comment (Maayan-s) [inline macros/edr/metadata_collection/get_tables_from_information_schema.sql:47]: same

File: macros/edr/metadata_collection/get_tables_from_information_schema.sql [hunk 13]
```diff
@@ -65,6 +62,27 @@
         schema_name,
         table_name
     from information_schema_tables
+{% endmacro %}
+
+{% macro postgres__get_tables_from_information_schema(schema_tuple) %}
+    {%- set database_name, schema_name = schema_tuple %}
+
+    with information_schema_tables as (
```
Comment (Maayan-s) [inline macros/edr/metadata_collection/get_tables_from_information_schema.sql:70]: Might be missing "(" on the beginning

File: macros/edr/tests/test_utils/get_compiled_test_id.sql (hunk 4 above)
Comment (haritamar) [reply macros/edr/tests/test_utils/get_compiled_test_id.sql:5]: Hmm I see, makes sense
Any chance you can give it a more direct name like "get_elementary_test_table_name"? (and avoid using it in logs etc)
Is it needed anywhere else?

Comment (haritamar) [reply macros/edr/tests/test_utils/get_compiled_test_id.sql:5]: Also - one problem I see with truncating from the end is that you might have a problematic character in the middle of a test name (e.g. a digit)
So you may need to add something at the beginning.

Or alternatively - truncate from the beginning but make sure the hash is added anyway.

File: macros/utils/data_types/cast_column.sql (hunk 5 above)
Comment (haritamar) [reply macros/utils/data_types/cast_column.sql:11]: 👍

File: macros/utils/table_operations/create_or_replace.sql (hunk 6 above)
Comment (haritamar) [reply macros/utils/table_operations/create_or_replace.sql:20]: OK got it
Didn't realize it was needed by the implementation

File: macros/edr/metadata_collection/get_tables_from_information_schema.sql (hunk 11 above)
Comment (haritamar) [reply macros/edr/metadata_collection/get_tables_from_information_schema.sql:10]: I think it was because of my comment (I didn't realize why it's needed)

Comment (elongl) [reply macros/edr/metadata_collection/get_tables_from_information_schema.sql:10]: Moved to the [macro itself](https://github.com/elementary-data/dbt-data-reliability/pull/230/commits/6fc9b19c4d3d0e0ed1afe972b8795cab44350d30).

File: macros/edr/tests/test_utils/get_compiled_test_id.sql (hunk 4 above)
Comment (elongl) [reply macros/edr/tests/test_utils/get_compiled_test_id.sql]: Thanks, that's great feedback.
I changed the implementation [here](https://github.com/elementary-data/dbt-data-reliability/pull/230/commits/90500cc2bc4c7a6bc3eb026dcdb3065e427d8972).
One thing that I'm not too crazy about is that singular tests don't have a hash at their unique ID.
Generally it doesn't matter because this macro is only called on Elementary's generic tests, but I wonder if we can find a different and more stable identifier for a test that is short. Ideally I'd hash the `unique_id` but I can't because `local_md5` is only introduced in v1.4.0. This might be good enough though.

**Review by haritamar — APPROVED** [review]
